src/main/scala/cognite/spark/v1/StreamIterator.scala

```diff
@@ -73,7 +73,7 @@ object StreamIterator {
     stream.chunks.parEvalMapUnordered(queueBufferSize) { chunk =>
       IO {
         val processedChunk = processChunk.map(f => f(chunk)).getOrElse(chunk)
-        queue.put(Right(processedChunk))
+        queue.offer(Right(processedChunk), 30, TimeUnit.SECONDS)
```

Comment:
Hm, I see. So all compute threads end up here, resulting in a deadlock, as no threads are available to do the reading from this queue?

If that's the case, I think the real fix might be to shift this to a different thread pool.

Reply:
I thought this would only be semantically blocking, not actually blocking the thread. :thinking:

Reply:
Oh, it's actually using `java.util.concurrent.ArrayBlockingQueue`. That's certainly thread blocking. Maybe it should be using a cats-effect `Queue` instead. Need to think about this for a bit.

Reply:
I think we're being totally stupid here, this should absolutely not be done on the cats-effect compute thread pool. :exploding_head:

Reply:
@vuhainguyen could  you try simply changing line 74 to `IO.blocking` (and no other changes) and testing that with the DMS test case you have? If this is the problem, I think it shouldn't hang any longer.

Reply:
https://typelevel.org/cats-effect/docs/thread-model#blocking
Are we already seeing a CE3 win here? :cat2:

Reply:
@hakontro yes, kind of! This is trivial to fix with CE3, in CE2 we would have had to create a thread pool manually and shifted and evaluated on and what not.

Reply:
Good news! From the article
> Notably, ContextShift and Blocker are no more.
Sounds like this is something we should look into, this isn't our first threadpool-related problem after all..